### DISCUSSION TITLE
Model Viewsets with no lookup field - Some hacks and some ideas.

### DISCUSSION THREAD

**Discussion by sergioisidoro** [encode/django-rest-framework#8305] [Ideas & Suggestions]
I'm working on an API endpoint which is a 1-to-1 with the current logged in user. I want this endpoint to work on a single object, which is the object associated with the current logged in user, like so:

```
GET api.acme.com/profiles/me/
POST api.acme.com/profiles/me/
PUT api.acme.com/profiles/me/

or like

GET api.acme.com/my-profile/
POST api.acme.com/my-profile/
PUT api.acme.com/my-profile/
``` 

However, routers really require a `lookupfield` because of the URL generation.  So I couldn't find any quick good approaches for this. I did come up with two hacks, which neither feel right to me:

### Using `as_view` + `get_object`
This means bypassing the router all together, and registering a path:
```
    path('my-profile',
         OwnProfileViewSet.as_view(
             {'get': 'retrieve', 'post': 'create', 'put': 'update', 'patch': 'partial_update'}
         ),
         name='my-profile'),
```
But this comes with some disadvantages, like automatic documentation, and other libraries that depend on the DRF router not working.

### Hacking the lookup field
This approach basically means creating a lookup field shortcut in the viewset, so when we pass the lookupfield with the value `me` it will return the object associated with the current user:

```
class ProfileViewSet(viewsets.ModelViewSet):
    ....
    def get_object(self):
        if self.kwargs['id'] == 'me':
            if self.request.user.is_anonymous:
                return None
            else:
                return self.request.user.profile
        else:
            return super(ProfileViewSet, self).get_object(........

```

### Discussion
I'm wondering what approaches other people used for these kinds of endpoints with DRF. 

* Did you just go with nested serialisers on the user?
* Did you make a custom router for these kinds of urls?
* I think it would be interesting to have the base router `register` method receive an optional`url` and the same arguments as `as_view` does, to register these kinds of endpoints like:
```
router.register_path("my-profile", OwnProfileViewSet, mappers={'get': 'retrieve', 'post': 'create', 'put': 'update', 'patch': 'partial_update'}
```

**Comment by ricky-sb** [comment]
How did you end up handling this?

**Comment by sergioisidoro** [comment]
@ricky-sb I ended going fully rest, and ditching these "current user" endpoints. The way I did it was just returning the user ID in the login response so the client knows the user ID. From there I just used went fully REST and serialising the attributes)
```
POST /login
{
   'auth_token': '....'
   'user_id': 1234
}

GET api/v1/users/1234
{
  'profile_id': _______
}
```


Another "hack" I did to facilitate things was tho make User and the related One to One model share the same primary key:
```
class UserProfile(models.Model):
    user = models.OneToOneField(
        'users.User',
        on_delete=models.CASCADE,
        primary_key=True,
    )
```

That way I get the user id in the login response, and I can do this at the same time in parallel without waiting for the user endpoint to return the ID of the profile associated with the user:
```
GET api/v1/users/1234
GET api/v1/profiles/1234
``` 

Hope this helps
